vdev/src/platform.rs

```diff
@@ -0,0 +1,50 @@
+use cached::proc_macro::once;
+use dunce;
+use home;
+use os_info;
+use std::path::{Path, PathBuf};
+
+#[once]
+fn _os_info() -> os_info::Info {
+    os_info::get()
+}
+
+pub struct Platform {}
+
+impl Platform {
+    pub fn new() -> Platform {
+        Platform {}
+    }
+
+    pub fn canonicalize_path(&self, path: &String) -> String {
+        match dunce::canonicalize(Path::new(&path)) {
+            Ok(p) => p.display().to_string(),
+            Err(_) => path.to_string(),
+        }
+    }
+
+    #[allow(dead_code)]
+    pub fn home(&self) -> PathBuf {
+        match home::home_dir() {
+            Some(path) => path,
+            None => ["~"].iter().collect(),
+        }
+    }
+
+    pub fn os_type(&self) -> os_info::Type {
+        _os_info().os_type()
+    }
+
+    pub fn windows(&self) -> bool {
+        self.os_type() == os_info::Type::Windows
+    }
+
+    pub fn macos(&self) -> bool {
+        self.os_type() == os_info::Type::Macos
+    }
+
+    #[allow(dead_code)]
+    pub fn linux(&self) -> bool {
+        !(self.windows() || self.macos())
+    }
```

Comment:
I think this OS type distinction is being made at build time, even in the `os_info` crate. Could this be achieved with `#[cfg(target_os = "X")]` instead of run-time checks?

Reply:
Sure! What would that look like? Also, is there any possibility that the values would differ?

Reply:
There are [two ways to go about this](https://doc.rust-lang.org/rust-by-example/attribute/cfg.html). As a first step, these could become (for example):
```rust
    pub const fn windows(&self) -> bool {
        #[cfg(target_os = "windows")]
        true
        #[cfg(not(target_os = "windows"))]
        false
    }
```
or simply:
```rust
    pub const fn windows(&self) -> bool {
        cfg!(target_os = "windows")
    }
```
However, since these now no longer depend on `self`, they could become struct methods called as `Platform::windows()`, or even just dropped and the `cfg` check moved into the callers directly.

The `cfg` target OS/family attributes are [documented in the reference book](https://doc.rust-lang.org/reference/conditional-compilation.html#target_os). We also use `#[cfg(windows)]` in a few places, which is equivalent to `target_family = "windows"`. We should probably standardize on one or the other.

As far as I am reading from the source code, the `os_info` crate is using those `cfg` attributes to generate at least the `os_type` return value, so I would not expect that value to differ. That crate provides more information than just the OS type, so there might be differences in other values.

Reply:
Thanks a lot! I pushed an update 👍